msprime/likelihood.py

```diff
@@ -0,0 +1,141 @@
+#
+# Copyright (C) 2019 University of Oxford
+#
+# This file is part of msprime.
+#
+# msprime is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# msprime is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Module responsible for computing likelihoods.
+"""
+import collections
+import math
+import msprime
+
+
+def unnormalised_log_mutation_likelihood(arg, theta):
+    # log_likelihood of mutations on a given ARG up to a normalising constant
+    # that depends on the pattern of observed mutations, but not on the ARG
+    # or the mutation rate
+    tables = arg.tables
+    edges_above = collections.defaultdict(list)
+    edges_below = collections.defaultdict(list)
+    total_material = 0
+    time = tables.nodes.time
+    for edge in tables.edges:
+        total_material += ((edge.right - edge.left) *
+                           (time[edge.parent] - time[edge.child]))
+        edges_above[edge.child].append(edge)
+        edges_below[edge.parent].append(edge)
+    number_of_mutations = len(tables.mutations)
+    if theta == 0:
+        if number_of_mutations == 0:
+            ret = 0
+        else:
+            ret = -float("inf")
+    else:
+        ret = (number_of_mutations * math.log(total_material * theta) -
+               total_material * theta)
+    for mut in arg.mutations():
+        mutant_location = tables.sites[mut.site].position
+        index = 0
+        to_check = edges_above[mut.node]
+        while to_check[index].right < mutant_location:
+            index += 1
+        edge = edges_above[mut.node][index]
+        potential_branch_length = time[edge.parent] - time[edge.child]
```

Comment:
Wouldn't it be simpler to compute this using the trees? We're basically iterating up and down the tree at the position for each mutation, aren't we? For each edge we examine, we're only looking at the branch length right, not the 'area' of the edge?

So,
```python
for tree in arg.trees():
    for site in tree.sites():
        assert len(site.mutations) == 1
        mutation = site.mutations[0]
        # Compute ``potential_branch_length`` by traversing up and down the tree?
        u = tree.parent(mutation.node)
        potential_branch_length = 0
        # Traverse up
        while u != tskit.NULL:
             potential_branch_length += tree.branch_length(u) # branch length of a root is 0
         # Traverse down (but we could do this without traversals if this is what we want)
         for u in tree.nodes(mutation.node):
             potential_branch_length += tree.branch_length(u)
```

Reply:
It is be possible to iterate over trees as well, and you're right that we only want the branch length, not the edge area. It's not quite as simple as your code though, because we still need to detect the start and end points of the window of valid of edges that are ancestral to exactly the right set of leaves at the mutant site. That window will typically not span the whole tree unless I'm dramatically misunderstanding the data structure. I can have a think about trying to write the function this way, and see if it makes a difference.

I had missed that `branch_length` is an implemented method. That will certainly make the code a little more readable as well.

Reply:
Ah, I see. This [issue](https://github.com/tskit-dev/tskit/issues/186) has some discussion on how to get edge information for tree nodes, which may be useful.

I think it's probably useful to recast this in terms of trees all right though since we are inherently doing tree traversals.